lib/parser/canonicalize-omp.cc

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "canonicalize-omp.h"
+#include "parse-tree-visitor.h"
+#include "parse-tree.h"
+
+// After Loop Canonicalization, rewrite OpenMP parse tree to make OpenMP
+// Constructs more structured which provide explicit scopes for later
+// structural checks and semantic analysis.
+//   1. move structured DoConstruct and OpenMPEndloopdirective into
+//      OpenMPLoopConstruct. Compilation will not proceed in case of errors
+//      after this pass.
+//   2. TODO: Start and End directive pair matching
+//   3. TBD
+namespace Fortran::parser {
+
+class CanonicalizationOfOmp {
+public:
+  template<typename T> bool Pre(T &) { return true; }
+  template<typename T> void Post(T &) {}
+  CanonicalizationOfOmp(Messages &messages) : messages_{messages} {}
+
+  void Post(Block &block) {
+    for (auto it{block.begin()}; it != block.end(); ++it) {
+      if (auto *exec{std::get_if<ExecutableConstruct>(&it->u)}) {
+        if (auto *ompCons{
+                std::get_if<common::Indirection<OpenMPConstruct>>(&exec->u)}) {
+          // OpenMPLoopConstruct
+          if (auto *ompLoop{
+                  std::get_if<OpenMPLoopConstruct>(&ompCons->value().u)}) {
+            RewriteOpenMPLoopConstruct(ompLoop, block, it);
+          }
+        } else if (auto *endDir{
+                       std::get_if<common::Indirection<OpenMPEndLoopDirective>>(
+                           &exec->u)}) {
+          // Unmatched OpenMPEndloopdirective
+          messages_.Say(endDir->value().source,
+              "The %s directive must follow the DO loop associated with the "
+              "loop construct"_err_en_US,
+              parser::ToUpperCaseLetters(endDir->value().source.ToString()));
+        }
+      }
+    }  // Block list
+  }
+
+private:
+  void RewriteOpenMPLoopConstruct(
+      OpenMPLoopConstruct *x, Block &block, Block::iterator it) {
+    // Check the sequence of DoConstruct and OpenMPEndLoopDirective
+    // in the same iteration
+    //
+    // Original:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //   ExecutableConstruct -> DoConstruct
+    //   ExecutableConstruct -> OpenMPEndLoopDirective (if available)
+    //
+    // After rewriting:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //     DoConstruct
+    //     OpenMPEndLoopDirective (if available)
+    Block::iterator nextIt;
+    OpenMPLoopConstruct *matchedLoopConstruct{nullptr};
+
+    auto &dir{std::get<OmpLoopDirective>(x->t)};
+    nextIt = it;
+    if (++nextIt != block.end()) {
+      if (auto *execNext{std::get_if<ExecutableConstruct>(&nextIt->u)}) {
+        if (auto *doCons{
+                std::get_if<common::Indirection<DoConstruct>>(&execNext->u)}) {
```

Comment:
This pattern repeats several times: check if you have an `ExecutableConstruct` and if so try to get a certain construct out of that. Here's is an example of a helper function that you should be able to use to simplify each of those places:
```
template<typename T> T *GetConstructIf(parser::ExecutionPartConstruct &x) {
  if (auto *y{std::get_if<ExecutableConstruct>(&x.u)}) {
    if (auto *z{std::get_if<common::Indirection<T>>(&y->u)}) {
      return &z->value();
    }
  }
  return nullptr;
}
```

Here, for example, you would replace the above two `if` statements with:
`if (auto *doCons{GetConstructIf<DoConstruct>(*nextIt)}) {`

Reply:
Thank you, I will add that!